augur/io.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+"""Interfaces for reading and writing data also known as input/output (I/O)
+"""
+import Bio.SeqIO
+import Bio.SeqRecord
+from pathlib import Path
+from xopen import xopen
+
+
+def read_sequences(paths, format):
+    """Read sequences from one or more paths.
+
+    Automatically infer compression mode (e.g., gzip, etc.) and return a stream
+    of sequence records in the requested format (e.g., "fasta", "genbank", etc.).
+
+    Parameters
+    ----------
+    paths : str, Path-like, or list of str or Path-like objects
+        One or more paths to sequence files of any type supported by BioPython.
+
+    format : str
+        Format of input sequences matching any of those supported by BioPython
+        (e.g., "fasta", "genbank", etc.).
+
+    Yields
+    ------
+    Bio.SeqRecord.SeqRecord
+        Sequence record from the given path(s).
+
+    """
+    try:
+        # Since we accept any Path-like string as an input, we try to cast the
+        # given path to a Path instance and then place it in a list. Note that
+        # this approach could break if we eventually support arbitrary URLs or
+        # paths to S3 buckets, etc.
+        paths = [Path(paths)]
+    except TypeError:
+        # If the given input is not a Path-like string, we assume it is an
+        # iterable sequence already.
+        pass
```

Comment:
That is the intended behavior, but I'm definitely not happy with this as a final implementation. What I was trying to do here was avoid explicitly testing for input types as lists or other iterables and instead use a duck-typing approach.

I always struggle with this question of how to properly do this in Python since so many valid potential types can be passed and explicitly testing for each one makes the interface more fragile when someone tries to pass a new valid type that you didn't consider.

I do think the `read_sequences` interface should support scalar or iterable inputs for the first argument, but we could implement that support differently to avoid this testing question. For example, we could use the star-arg interface and make `format` a named argument like this:

```
def read_sequences(*paths, format=None):
    pass
```

This interface puts the burden of passing paths into the function on the caller to do the right thing like:

```
# Pass multiple paths...
paths = ["foo", "bar"]
read_sequences(*paths, format="fasta")

# Or pass a single path:
read_sequences("foo", format="fasta")
```

Or we could just require the input to always be an iterable.